File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.go
func search(arr []int, target int) int {
    start, end := 0, len(arr)-1
    
    for start <= end {
        mid := start + (end-start)/2
        
        // Element found
        if arr[mid] == target {
            return mid
        }
        
        // Left side is sorted
        if arr[mid] >= arr[0] {
            if arr[start] <= target && target <= arr[mid] {
                end = mid - 1
            } else {
                start = mid + 1
            }
        } else { 
            // Right side is sorted
            if arr[mid] <= target && target <= arr[end] {
                start = mid + 1
            } else {
                end = mid - 1
            }
        }
    }
    
    return -1
}

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums,target):
        start,end=0,len(nums)-1
        
        while start<=end:
            mid=start+(end-start)//2

            # find element
            if nums[mid]==target:
                return mid
                # left side sorted 
            elif nums[mid]>=nums[0]:
                if nums[start]<=target<=nums[mid]:
                    end=mid-1
                else:
                    start=mid+1
            # right side sorted
            else:
                if nums[mid]<=target<=nums[end]:
                    start=mid+1
                else:
                    end=mid-1
        return -1